File: leetcode/201-300/243-shortest-word-distance/bench/shortest_distance.c
/* Benchmark harness for LeetCode #243 — Shortest Word Distance.
 * Mirrors shortest_distance.kara algorithm-for-algorithm.
 *
 * On the string representation — this is the one place a C mirror can quietly
 * stop measuring the same operation as the other four. C's reflex here is
 * `strcmp`, which carries no length and must walk to a NUL or a mismatch. The
 * other four languages all hold (pointer, length) and compare the length first.
 * A strcmp mirror would therefore be benchmarking a DIFFERENT primitive, and
 * since every word here is the same 9 bytes long the length check never fires
 * anyway — the difference would be pure representation overhead, not algorithm.
 *
 * So `Str` carries an explicit length and `str_eq` does length-then-memcmp,
 * exactly what Rust's `String == &str`, Go's `==` and kāra's `String ==` do.
 *
 * Each of the 20,000 slots gets its OWN copy of its word (its own row of
 * `pool`), matching the `.clone()` in the kāra and Rust mirrors, so no lane can
 * shortcut an equality test by noticing the operands share a data pointer.
 */

#include <stdio.h>
#include <string.h>
#include <stdint.h>

#define VOCAB_N 256
#define N 20000
#define ITERS 2000
#define WLEN 9

typedef struct {
    const char *p;
    int32_t len;
} Str;

static inline int str_eq(Str a, Str b) {
    return a.len == b.len && memcmp(a.p, b.p, (size_t)a.len) == 0;
}

static long long shortest_distance(const Str *words, long long n, Str w1, Str w2) {
    long long last1 = -1, last2 = -1;
    long long best = n;
    for (long long i = 0; i < n; i++) {
        if (str_eq(words[i], w1)) {
            last1 = i;
            if (last2 >= 0) {
                long long d = last1 - last2;
                best = d < best ? d : best;
            }
        } else if (str_eq(words[i], w2)) {
            last2 = i;
            if (last1 >= 0) {
                long long d = last2 - last1;
                best = d < best ? d : best;
            }
        }
    }
    return best;
}

/* Overflow-free 31-bit LCG; every draw uses bits 16..23. */
static long long lcg(long long state) {
    return (state * 1103515245LL + 12345LL) & 2147483647LL;
}

static char vocab_buf[VOCAB_N][WLEN + 1];
static Str vocab[VOCAB_N];
static char pool[N][WLEN + 1];
static Str list[N];

int main(void) {
    const char alpha[4] = {'a', 'b', 'c', 'd'};

    for (long long v = 0; v < VOCAB_N; v++) {
        memcpy(vocab_buf[v], "delta", 5);
        vocab_buf[v][5] = alpha[(v / 64) % 4];
        vocab_buf[v][6] = alpha[(v / 16) % 4];
        vocab_buf[v][7] = alpha[(v / 4) % 4];
        vocab_buf[v][8] = alpha[v % 4];
        vocab_buf[v][9] = '\0';
        vocab[v].p = vocab_buf[v];
        vocab[v].len = WLEN;
    }

    long long state = 1;
    for (long long i = 0; i < N; i++) {
        state = lcg(state);
        long long j = (state / 65536) % VOCAB_N;
        memcpy(pool[i], vocab_buf[j], WLEN + 1);
        list[i].p = pool[i];
        list[i].len = WLEN;
    }

    long long acc = 0;
    long long qstate = 7;
    for (long long k = 0; k < ITERS; k++) {
        qstate = lcg(qstate);
        long long a = (qstate / 65536) % VOCAB_N;
        qstate = lcg(qstate);
        long long b = (qstate / 65536) % VOCAB_N;
        if (b == a) {
            b = (b + 1) % VOCAB_N;
        }
        long long d = shortest_distance(list, N, vocab[a], vocab[b]);
        acc = (acc * 131 + d) % 1000000007LL;
    }
    printf("%lld\n", acc);
    return 0;
}

File: leetcode/201-300/243-shortest-word-distance/bench/shortest_distance.rs
// Benchmark harness for LeetCode #243 — Shortest Word Distance.
// Mirrors shortest_distance.kara algorithm-for-algorithm.
//
// `.clone()` per slot is deliberate and matches the kāra mirror: each of the
// 20,000 slots owns its word, so no equality test can shortcut on the two
// operands sharing a data pointer.

fn shortest_distance(words: &[String], word1: &str, word2: &str) -> i64 {
    let n = words.len() as i64;
    let mut last1: i64 = -1;
    let mut last2: i64 = -1;
    let mut best = n;
    let mut i: i64 = 0;
    while i < n {
        if words[i as usize] == word1 {
            last1 = i;
            if last2 >= 0 {
                best = best.min(last1 - last2);
            }
        } else if words[i as usize] == word2 {
            last2 = i;
            if last1 >= 0 {
                best = best.min(last2 - last1);
            }
        }
        i += 1;
    }
    best
}

// Overflow-free 31-bit LCG; every draw uses bits 16..23.
fn lcg(state: i64) -> i64 {
    (state * 1103515245 + 12345) & 2147483647
}

fn main() {
    let vocab_n: i64 = 256;
    let n: i64 = 20000;
    let iters: i64 = 2000;

    let alpha = ["a", "b", "c", "d"];
    let mut vocab: Vec<String> = Vec::new();
    let mut v: i64 = 0;
    while v < vocab_n {
        let mut w = String::from("delta");
        w.push_str(alpha[((v / 64) % 4) as usize]);
        w.push_str(alpha[((v / 16) % 4) as usize]);
        w.push_str(alpha[((v / 4) % 4) as usize]);
        w.push_str(alpha[(v % 4) as usize]);
        vocab.push(w);
        v += 1;
    }

    let mut list: Vec<String> = Vec::new();
    let mut state: i64 = 1;
    let mut i: i64 = 0;
    while i < n {
        state = lcg(state);
        list.push(vocab[((state / 65536) % vocab_n) as usize].clone());
        i += 1;
    }

    let mut acc: i64 = 0;
    let mut qstate: i64 = 7;
    let mut k: i64 = 0;
    while k < iters {
        qstate = lcg(qstate);
        let a = (qstate / 65536) % vocab_n;
        qstate = lcg(qstate);
        let mut b = (qstate / 65536) % vocab_n;
        if b == a {
            b = (b + 1) % vocab_n;
        }
        let d = shortest_distance(&list, &vocab[a as usize], &vocab[b as usize]);
        acc = (acc * 131 + d) % 1000000007;
        k += 1;
    }
    println!("{}", acc);
}
